samcli/lib/build/app_builder.py

```diff
@@ -0,0 +1,305 @@
+                                     runtime):
+
+        # If we are printing debug logs in SAM CLI, the builder library should also print debug logs
+        log_level = LOG.getEffectiveLevel()
+
+        container = LambdaBuildContainer(lambda_builders_protocol_version,
+                                         config.language,
+                                         config.dependency_manager,
+                                         config.application_framework,
+                                         source_dir,
+                                         manifest_path,
+                                         runtime,
+                                         log_level=log_level,
+                                         optimizations=None,
+                                         options=None)
+
+        try:
+            self._container_manager.run(container)
+        except docker.errors.APIError as ex:
+            if "executable file not found in $PATH" in str(ex):
+                raise UnsupportedBuilderLibraryVersionError(container.image,
+                                                            "{} executable not found in container"
+                                                            .format(container.executable_name))
+
+        # Container's output provides status of whether the build succeeded or failed
+        # stdout contains the result of JSON-RPC call
+        stdout_stream = io.BytesIO()
+        # stderr contains logs printed by the builder. Stream it directly to terminal
+        stderr_stream = osutils.stderr()
+        container.wait_for_logs(stdout=stdout_stream, stderr=stderr_stream)
+
+        stdout_data = stdout_stream.getvalue().decode('utf-8')
+        LOG.debug("Build inside container returned response %s", stdout_data)
+
+        response = self._parse_builder_response(stdout_data, container.image)
+
+        # Request is successful. Now copy the artifacts back to the host
+        LOG.debug("Build inside container was successful. Copying artifacts from container to host")
+
+        # "/." is a Docker thing that instructions the copy command to download contents of the folder only
+        result_dir_in_container = response["result"]["artifacts_dir"] + "/."
+        container.copy(result_dir_in_container, artifacts_dir)
+
+        LOG.debug("Build inside container succeeded")
+        return artifacts_dir
+
+    @staticmethod
+    def _parse_builder_response(stdout_data, image_name):
+
+        try:
+            response = json.loads(stdout_data)
+        except Exception:
+            # Invalid JSON is produced as an output only when the builder process crashed for some reason.
+            # Report this as a crash
+            LOG.debug("Builder crashed")
+            raise
+
+        if "error" in response:
+            error = response.get("error", {})
+            err_code = error.get("code")
+            msg = error.get("message")
+
+            if 400 <= err_code < 500:
```

Comment:
errorcode to exception mapping, can be encapsulated in a dict as well.

```
{ 
"505':UnspoortedError,
"-400": BuildError
}
```

Helps avoid code branching logic using just if statements.

Reply:
oh yeah! this will be really neat. I am going to stick with if/else for now because it offers more flexibility